pkg/cloud/services/ec2/subnets.go

```diff
@@ -54,36 +54,49 @@ func (s *Service) reconcileSubnets() error {
 		return err
 	}
 
-	// If the subnets are empty, populate the slice with the default configuration.
-	// Adds a single private and public subnet in the first available zone.
-	if len(existing) < 2 && len(subnets) < 2 {
-		zones, err := s.getAvailableZones()
-		if err != nil {
-			return err
+	// If we have an unmanaged VPC then we expect the following to be true:
+	// 1) Subnets must be specified
+	// 2) The subnets specified must already exist
+	// Otherwise this is an error
```

Comment:
This seems like a pretty significant change to me – previously CAPA would "discover" the subnets attached to a VPC and whether they were private/public. With this change, it looks to me like it still will, as long as the spec is "seeded" with a single public or private subnet (see https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/1721/files#diff-b636b6a430a5b4f76ee6d31bc48bf46aR112-R116 ).

I'm not sure I follow how this change is important for EKS support, would you mind explaining?

Reply:
This part of the change isn't important to EKS support. This part of the change was more around making the 'unmanaged vpc' so that you need to explicitly create and specify your subnets....properly bring your own subnets.

The important part for the EKS support is that the default subnetting strategy for a managed vpc is multi-az. Although its not just for EKS, it feels better that the default subnet strategy is multi-az by default across the board.

An additional check will need to be added in this area in the future for unmanaged vpc's to check that there are subnets in 2 azs but i haven't added this yet as its only a requirment for EKS.

Reply:
Oh, I see. I think this belongs much further down in the function, then, after CAPA has discovered the existing subnets and populated metadata (critically: whether they're public or private). We're going to overwrite what the user provided anyway with what we discovered, so it would be better to validate against the observed state than what they specified.

Also, would you mind [adding an event to the AWSCluster object](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/6999f6dd5d41f63e7b783d3676bd6b484a31e391/pkg/cloud/services/ec2/gateways.go#L123) so we have a description of the problem when someone does a `k describe awscluster FOO`?

Reply:
Thanks @sethp-nr , i had missed that part about updating the subnets in the spec with the details of the existing subnets. As you pointed out, critical for the public/private test. I have updated this and also refactored the code further. And i've also added events as requested.